### PR TITLE
[TASK] Performance optimisation, test environment & toolchain

### PR DESCRIPTION
**Pull request by teneris**
## Summary

This PR adds a complete development toolchain, performance optimisations, caching improvements, deprecation fixes, and CI/release automation for the pagebased extension.

---

### 🛠 Development Toolchain
- DDEV environment (PHP 8.3, MariaDB 10.11) with `ddev init`, `ddev test`, `ddev seed` commands
- Composer scripts: `lint`, `fix`, `sca`, `test`, `test:coverage`
- php-cs-fixer, PHPStan (level 8 + baseline), Rector, PHPMD, EditorConfig
- CaptainHook: commit-msg format validation, pre-commit lint/SCA, pre-push tests

### ✅ Tests
- **60 unit tests**: Pagination, CastUtility, TagUtility
- **39 functional tests**: RootLineUtility, AbstractObjectRepository, performance/query-count tests
- `QueryCountingMiddleware` for verifying DB query counts via Doctrine DBAL middleware

### ⚡ Performance Fixes
- **DB indexes** on `_pagebased_registration`, `_pagebased_site`, composite index on both
- **RootLineUtility**: request-scoped static cache prevents N+1 recursive DB queries
- **`findByUid()`**: minimal 2-constraint direct query instead of full demand pipeline
- **TagUtility**: O(1) hash map deduplication (was O(n²) `in_array`)
- **`findTagStrings()`**: direct `SELECT` of tag column only — no Extbase object hydration
- **RegistrationService**: 7 static lookup indices, all `getRegistrationBy*()` now O(1)

### 🗄 Caching
- TYPO3 `pagebased_tags` cache (VariableFrontend, 24h, group: `pages`) for tag queries
- Cache auto-invalidated on page save via TYPO3's page cache group

### 🔧 Deprecation Fixes
- All 6 `QueryBuilder::execute()` calls replaced with `executeQuery()` (TYPO3 v13 compatibility)

### 🚀 CI / Release
- GitHub Actions CI: lint + tests matrix (PHP 8.3 & 8.4) using `ddev/github-action-setup-ddev`
- GitHub Actions Release: auto-detects semver bump from TYPO3 commit prefixes, generates changelog via git-cliff, creates annotated tag + GitHub Release (dry-run mode available)

---

## No Breaking Changes

All changes are additive or internal. Public APIs and extension registration behaviour are unchanged.

### CODE REVIEW

File: Classes/Domain/Repository/AbstractObjectRepository.php [hunk 1]
```diff
@@ -96,6 +100,93 @@ public function createDemandConstraints(DemandInterface $demand, QueryInterface
...
+            );
+
+        if ($categoryUid > 0) {
+            $pageIds = array_keys(RootLineUtility::collectPagesBelow($categoryUid));
+            if (!empty($pageIds)) {
+                $qb->andWhere($qb->expr()->in('uid', $pageIds));
+            } else {
+                return [];
+            }
+        }
```
[the remark this answers is not in the document]
Comment (teneris) [reply Classes/Domain/Repository/AbstractObjectRepository.php:173]: Fixed: array binding now uses `$qb->createNamedParameter($pageIds, Connection::PARAM_INT_ARRAY)`, which generates valid SQL via Doctrine DBAL's typed array parameter binding.

File: Classes/Domain/Repository/AbstractObjectRepository.php [hunk 2]
```diff
@@ -96,6 +100,93 @@ public function createDemandConstraints(DemandInterface $demand, QueryInterface
...
+
+        $result = array_column($qb->executeQuery()->fetchAllAssociative(), 'pagebased_tags');
+
+        $cache?->set($cacheKey, $result, ['pageId_1']);
```
[the remark this answers is not in the document]
Comment (teneris) [reply Classes/Domain/Repository/AbstractObjectRepository.php]: Fixed in a previous commit: cache tags are now derived dynamically as `array_map(static fn($row) => 'pageId_' . $row['uid'], $rows)`, so TYPO3 invalidates the correct page UIDs instead of the hardcoded `pageId_1`.

File: Classes/Domain/Repository/AbstractObjectRepository.php [hunk 3]
```diff
@@ -96,6 +100,93 @@ public function createDemandConstraints(DemandInterface $demand, QueryInterface
...
+     * @param ObjectDemandInterface $demand Used for optional category-tree filtering.
+     * @return string[] Raw comma-separated tag strings, one entry per page row.
+     */
+    public function findTagStrings(ObjectDemandInterface $demand): array
+    {
+        $categoryUid = $demand->getCategory();
+        $languageUid = (int)GeneralUtility::makeInstance(Context::class)
+            ->getPropertyFromAspect('language', 'id', 0);
+        $cacheKey = 'pagebased_tags_' . md5(
+            $this->registration->getIdentifier() . '_' . $categoryUid . '_' . $languageUid
```
[the remark this answers is not in the document]
Comment (teneris) [reply Classes/Domain/Repository/AbstractObjectRepository.php:141]: Fixed: removed `$languageUid` from the cache key. Since `findTagStrings()` issues no language constraint in its query, the result is language-agnostic and the key must not vary by language context.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+          git config user.email "github-actions[bot]@users.noreply.github.com"
+          git add CHANGELOG.md
+          git diff --staged --quiet || git commit -m "[TASK] Update CHANGELOG.md for ${{ steps.version.outputs.version }}"
+
+      - name: Create tag
+        if: ${{ !inputs.dry_run }}
+        run: |
+          git tag -a "${{ steps.version.outputs.version }}" \
+            -m "Release ${{ steps.version.outputs.version }}"
+          git push origin HEAD "${{ steps.version.outputs.version }}"
```
[the remark this answers is not in the document]
Comment (teneris) [reply .github/workflows/release.yml:127]: Fixed: split into `git push origin "HEAD:${{ github.ref_name }}"` (explicit branch ref) and `git push origin "${{ steps.version.outputs.version }}"` (tag only), avoiding failures in detached HEAD state.

File: .ddev/config.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
+name: pagebased
+type: typo3
+docroot: .build/public
+php_version: "8.3"
```
[the remark this answers is not in the document]
Comment (teneris) [reply .ddev/config.yaml:4]: Fixed: changed `php8.3-pcov` to the version-agnostic `php-pcov`, which installs correctly regardless of the PHP version used in CI.

File: .ddev/config.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+router_http_port: "80"
+router_https_port: "443"
+xdebug_enabled: false
+additional_hostnames: []
+additional_fqdns: []
+database:
+  type: mariadb
+  version: "10.11"
+nodejs_version: "20"
+webimage_extra_packages: [php8.3-pcov]
```
[the remark this answers is not in the document]
Comment (teneris) [reply .ddev/config.yaml:15]: Fixed: changed `php8.3-pcov` to the version-agnostic `php-pcov`, which installs correctly regardless of the PHP version used in CI.

File: Classes/Domain/Repository/AbstractObjectRepository.php [hunk 7]
```diff
@@ -96,6 +100,93 @@ public function createDemandConstraints(DemandInterface $demand, QueryInterface
...
+    /**
+     * Fetches only the raw tag CSV strings from the pages table for this registration,
+     * bypassing full Extbase object hydration. Results are cached in TYPO3's data
+     * cache and automatically invalidated when pages are modified.
+     *
+     * @param ObjectDemandInterface $demand Used for optional category-tree filtering.
+     * @return string[] Raw comma-separated tag strings, one entry per page row.
+     */
+    public function findTagStrings(ObjectDemandInterface $demand): array
+    {
```
[the remark this answers is not in the document]
Comment (teneris) [reply Classes/Domain/Repository/AbstractObjectRepository.php:136]: Fixed: added `FindTagStringsTest.php` with 7 functional tests covering return type/count correctness, category filtering, the empty-category edge case, and cache-hit consistency across consecutive calls.

File: Classes/Domain/Repository/AbstractObjectRepository.php [hunk 8]
```diff
@@ -96,6 +101,95 @@ public function createDemandConstraints(DemandInterface $demand, QueryInterface
...
+            );
+
+        if ($categoryUid > 0) {
+            $pageIds = array_keys(RootLineUtility::collectPagesBelow($categoryUid));
+            if (!empty($pageIds)) {
+                $qb->andWhere($qb->expr()->in('uid', $qb->createNamedParameter($pageIds, Connection::PARAM_INT_ARRAY)));
+            } else {
+                return [];
+            }
+        }
```
[the remark this answers is not in the document]
Comment (teneris) [reply Classes/Domain/Repository/AbstractObjectRepository.php:173]: Fixed: `findTagStrings()` now replicates the same visibility constraints as the Extbase demand pipeline — `nav_hide = 0`, the `l18n_cfg` language-configuration constraint from `AbstractPageRepository`, and child-object exclusion when `$demand->getIncludeChildObjects() === false`. The `includeChildObjects` flag is also added to the cache key so different demand configurations are cached separately. TYPO3's QueryBuilder restriction container continues to handle `hidden`/`deleted`/`starttime`/`endtime` automatically.

File: Classes/Utility/TagUtility.php [hunk 9]
```diff
@@ -41,7 +64,15 @@ public static function getTags(ObjectDemandInterface $demand, RepositoryInterfac
...
         }
 
-        // Find objects and return their tags
+        // Use a lightweight tag-only query when supported (avoids full Extbase object hydration)
+        if ($repository instanceof AbstractObjectRepository) {
+            $tagDemand = $ignoreTagsFromDemand === true ? $demand->setTags(null) : $demand;
+            $tagStrings = $repository->findTagStrings($tagDemand);
+
+            return $tagStrings !== [] ? self::collectTagsFromStrings($tagStrings) : null;
+        }
```
[the remark this answers is not in the document]
Comment (teneris) [reply Classes/Utility/TagUtility.php:73]: Fixed: `getTags()` now skips the `findTagStrings()` fast path when `$languageUid \!== null`, falling back to the hydrated Extbase path which correctly applies language overlays and respects `setRespectSysLanguage()`. The lightweight path is only used for the default-language (frontend) case where language overlays are not needed.

File: .ddev/commands/web/test [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+## Usage: test [flags]
+## Example: "ddev test" or "ddev test --unit" or "ddev test --functional" or "ddev test --coverage"
+## Flags: [{"Name":"unit","Shorthand":"u","Usage":"Run unit tests only"},{"Name":"functional","Shorthand":"f","Usage":"Run functional tests only"},{"Name":"coverage","Shorthand":"c","Usage":"Run tests with coverage report"},{"Name":"filter","Usage":"PHPUnit filter expression"}]
+
+set -Eeuo pipefail
+
+# Auto-proxy into DDEV if running outside the container
+if [ -z "${IS_DDEV_PROJECT:-}" ]; then
+    if command -v ddev >/dev/null 2>&1 && [ -d ".ddev" ]; then
+        ddev exec bash -lc "/var/www/html/.ddev/commands/web/test $*"
```
[the remark this answers is not in the document]
Comment (teneris) [reply .ddev/commands/web/test:13]: Fixed: changed to `ddev exec bash -lc '/var/www/html/.ddev/commands/web/test "$@"' -- test "$@"`, which passes arguments via `$@` inside a single-quoted string with a separate `--` separator so all flags (including those with spaces or special characters like `--filter='some test'`) are forwarded correctly.

File: Tests/Functional/Performance/RootLineUtilityPerformanceTest.php [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+        // Warm call
+        $start = microtime(true);
+        RootLineUtility::collectPagesBelow(200);
+        $warmMs = (microtime(true) - $start) * 1000;
+
+        self::assertLessThan($coldMs, $warmMs, sprintf(
+            'Cached call (%.3f ms) should be faster than cold call (%.3f ms)',
+            $warmMs,
+            $coldMs
+        ));
```
[the remark this answers is not in the document]
Comment (teneris) [reply Tests/Functional/Performance/RootLineUtilityPerformanceTest.php:181]: Fixed: changed the assertion from strict `assertLessThan($coldMs, $warmMs)` to `assertLessThan($coldMs * 5, $warmMs)`, matching the documented 5× tolerance. Also updated the comment and error message to accurately describe what is being tested. The zero-query guarantee (the real correctness check) remains in the separate `collectPagesBelowIssuesZeroQueriesOnCacheHit()` test.

File: Tests/Functional/Performance/RepositoryPerformanceTest.php [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+        QueryCountingMiddleware::reset();
+
+        $this->repository->findByUid(500);
+
+        $queryCount = QueryCountingMiddleware::getCount();
+
+        self::assertLessThanOrEqual(3, $queryCount, sprintf(
+            'findByUid() should issue at most 3 queries, got %d',
+            $queryCount
+        ));
```
[the remark this answers is not in the document]
Comment (teneris) [reply Tests/Functional/Performance/RepositoryPerformanceTest.php:138]: Fixed: updated the docblock to say "at most 3 queries" to align with the `assertLessThanOrEqual(3, ...)` assertion.

File: Classes/Domain/Repository/AbstractObjectRepository.php [hunk 13]
```diff
@@ -96,6 +101,95 @@ public function createDemandConstraints(DemandInterface $demand, QueryInterface
...
+
+        $rows = $qb->executeQuery()->fetchAllAssociative();
+        $result = array_column($rows, 'pagebased_tags');
+
+        $cacheTags = array_map(static fn(array $row) => 'pageId_' . $row['uid'], $rows);
+        $cache?->set($cacheKey, $result, $cacheTags ?: ['pagebased_tags']);
+
+        return $result;
+    }
```
[the remark this answers is not in the document]
Comment (teneris) [reply Classes/Domain/Repository/AbstractObjectRepository.php:182]: Already addressed: `FindTagStringsTest.php` was added in a previous commit with 7 functional tests covering return correctness, category filtering, the empty-category edge case, and cache-hit consistency. Additionally, the demand-constraint gap noted in the related thread has now been fixed — `findTagStrings()` replicates `nav_hide`, `l18n_cfg`, and child-object exclusion constraints from the Extbase pipeline.

File: Classes/Domain/Repository/AbstractObjectRepository.php [hunk 14]
```diff
@@ -96,6 +101,95 @@ public function createDemandConstraints(DemandInterface $demand, QueryInterface
...
+        $categoryUid = $demand->getCategory();
+        $languageUid = (int)GeneralUtility::makeInstance(Context::class)
+            ->getPropertyFromAspect('language', 'id', 0);
+        $cacheKey = 'pagebased_tags_' . md5(
+            $this->registration->getIdentifier() . '_' . $categoryUid . '_' . $languageUid
+        );
+
+        $cache = $this->getTagsCache();
+        if ($cache !== null && ($cached = $cache->get($cacheKey)) !== false) {
+            return $cached;
```
[the remark this answers is not in the document]
Comment (teneris) [reply Classes/Domain/Repository/AbstractObjectRepository.php:146]: Already addressed in a previous commit: `$languageUid` was removed from the cache key since `findTagStrings()` issues no language constraint. Additionally, `getTags()` now skips `findTagStrings()` entirely when a non-null `$languageUid` is passed, so language-scoped callers always use the hydrated Extbase path.

**Comment by pdaether** [comment]
LGTM

**Review by pdaether — APPROVED** [review]
LGTM
